### PR TITLE
Rename Webpacker instances to shakapacker

### LINKED ISSUE #157: Rename all references of webpacker to shakapacker
This is the first item for v7.

- [x] Can we alias the constant Webpacker to Shakapacker? Should we display deprecation warnings?
- [x] Changelog
- [x] Tests and CI configuration
- [x] JS Code
- [x] Ruby Code
- [ ] Examples
- [x] Docs
- [x] Error messages
- [ ] Comments
- [ ] Update key example projects
- [ ] Effect of libraries
  - [ ] [React on Rails](https://github.com/shakacode/react_on_rails)
  - [ ] [React Rails](https://github.com/reactjs/react-rails)
- [ ] Generator effects for new app

### PR DESCRIPTION
**Pull request by ahangarha**
This PR should cover the following with backward compatibility:
- [x] Use Shakapacker module by default.
- [x] Use config files with names related to Shakapacker (namely `config/shakapacker.yml`)
- [x] Use Rake tasks with the `shakapacker` namespace.
- [x] Update tests
- [x] Update the documentation.
- [x] Guideline for migrating to Shakapacker naming conventions.
- [x] Update Changelog

Closes #157

### CODE REVIEW

File: lib/shakapacker/base_strategy.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module Shakapacker
```
Comment (vaukalak) [inline lib/shakapacker/base_strategy.rb:1]: @ahangarha I'm not sure if that's possible, but if this is just a copy of existent  `webpacker/` files, could we use `git-move`, to make sure we preserve the history?

Comment (vaukalak) [reply lib/shakapacker/base_strategy.rb:1]: I mean hear and other files.

File: lib/tasks/webpacker/check_node.rake [hunk 2]
```diff
@@ -1,31 +1,7 @@
...
-        $stderr.puts "Please upgrade Node.js https://nodejs.org/en/download/"
-        $stderr.puts "Exiting!"
-        exit!
-      end
-    rescue Errno::ENOENT
-      $stderr.puts "Node.js not installed. Please download and install Node.js https://nodejs.org/en/download/"
-      $stderr.puts "Exiting!"
-      exit!
-    end
+  desc "[DEPRECATED] Verifies if Node.js is installed"
```
Comment (vaukalak) [inline lib/tasks/webpacker/check_node.rake:2]: `[DEPRECATED] (user shakapacker...) Verifies if Node.js is installed`

File: lib/tasks/webpacker/info.rake [hunk 3]
```diff
@@ -1,21 +1,7 @@
...
-      $stdout.puts "Yarn: #{`yarn --version`}"
-
-      $stdout.puts "\n"
-      $stdout.puts "shakapacker: \n#{`npm list shakapacker version`}"
-
-      $stdout.puts "Is bin/webpacker present?: #{File.exist? 'bin/webpacker'}"
-      $stdout.puts "Is bin/webpacker-dev-server present?: #{File.exist? 'bin/webpacker-dev-server'}"
-      $stdout.puts "Is bin/yarn present?: #{File.exist? 'bin/yarn'}"
-    end
+  desc "[DEPRECATED] Provide information on Shakapacker's environment"
```
Comment (vaukalak) [inline lib/tasks/webpacker/info.rake:2]: I would add `[DEPRECATED] (user shakapacker...) Verifies if Node.js is installed` to all of this commands.

File: lib/shakapacker/base_strategy.rb (hunk 1 above)
Comment (ahangarha) [reply lib/shakapacker/base_strategy.rb:1]: I am not sure why but some files have proper history and some don't. For example [./lib/shakapacker/webpack_runner.rb](https://github.com/shakacode/shakapacker/blame/webpacker-to-shakapacker/lib/shakapacker/webpack_runner.rb) has proper commit history.

Any idea how to fix this?

File: lib/tasks/webpacker/check_node.rake (hunk 2 above)
Comment (ahangarha) [reply lib/tasks/webpacker/check_node.rake:2]: Initially, I was adding more information in all places. Then I thought maybe this is not necessary. I think as we go forward, we get better sense on where and how to show deprecatio warning.

File: lib/shakapacker.rb [hunk 4]
```diff
@@ -1 +1,61 @@
-require "webpacker"
+require "active_support/core_ext/module/attribute_accessors"
```
Comment (justin808) [inline lib/shakapacker.rb:1]: https://git-scm.com/docs/git-mv

File: lib/tasks/shakapacker.rake [hunk 5]
```diff
@@ -0,0 +1,18 @@
+tasks = {
+  "shakapacker:info"                    => "Provides information on Shakapacker's environment",
+  "shakapacker:install"                 => "Installs and setup webpack with Yarn",
+  "shakapacker:compile"                 => "Compiles webpack bundles based on environment",
+  "shakapacker:clean"                   => "Remove old compiled webpacks",
```
Comment (justin808) [inline lib/tasks/shakapacker.rake:5]: old compiled bundles

File: lib/tasks/webpacker/clean.rake [hunk 6]
```diff
@@ -1,23 +1,7 @@
...
-  if Rake::Task.task_defined?("assets:clean")
-    Rake::Task["assets:clean"].enhance do
-      Rake::Task["webpacker:clean"].invoke
-    end
-  else
-    Rake::Task.define_task("assets:clean" => "webpacker:clean")
+  desc "[DEPRECATED] Remove old compiled webpacks"
+  task :clean, [:keep, :age] do |_, args|
+    prefix = task.name.split(/#|webpacker:/).first
+    Rake::Task["#{prefix}shakapacker:clean"].invoke(args.keep, args.age)
```
Comment (justin808) [inline lib/tasks/webpacker/clean.rake:5]: alert deprecated before you call this

Comment (justin808) [reply lib/tasks/webpacker/clean.rake:5]: We can have another PR that adds back the deprecated rakes.

File: lib/webpacker/dev_server_runner.rb [hunk 7]
```diff
@@ -1,96 +1,15 @@
...
-        server = TCPServer.new(@hostname, @port)
-        server.close
-
-      rescue Errno::EADDRINUSE
-        $stdout.puts "Another program is running on port #{@port}. Set a new port in #{@config.config_path} for dev_server"
-        exit!
-      end
-
       def execute_cmd
         env = Webpacker::Compiler.env
```
Comment (justin808) [inline lib/webpacker/dev_server_runner.rb:11]: Why not Shakapacker::Compiler.env?

File: lib/webpacker/helper.rb [hunk 8]
```diff
@@ -1,160 +1,10 @@
 module Webpacker::Helper
-  # Returns the current Webpacker instance.
-  # Could be overridden to use multiple Webpacker
-  # configurations within the same app (e.g. with engines).
+  include Shakapacker::Helper
+
   def current_webpacker_instance
     Webpacker.instance
```
Comment (justin808) [inline lib/webpacker/helper.rb:5]: shakapacker.instance?

**Review by justin808 — CHANGES_REQUESTED** [review]
https://www.rubyguides.com/2018/10/delegate-methods-in-ruby/

PR should be much smaller.

File: lib/shakapacker.rb (hunk 4 above)
Comment (ahangarha) [reply lib/shakapacker.rb:1]: I did almsot the same thing in [this commit](https://github.com/shakacode/shakapacker/commit/45c67fca213f0ccb7ca108d06bdcb161d247c22b) but no effect.

There are two exceptions: one is this file (I could do better) and `railtie.rb` (where I don't know why it didn't consider it as a move). Rest should be fine but it is not.

File: lib/webpacker/dev_server_runner.rb (hunk 7 above)
Comment (ahangarha) [reply lib/webpacker/dev_server_runner.rb:11]: Just for having consistence in the code base. Still, `Webpacker::Comiler` inherits entirely from `Shakapacker::Compiler`: https://github.com/shakacode/shakapacker/blob/d3d84a02c9c565f89a74f9b5e30d1446c6d0b745/lib/webpacker/compiler.rb#L1

File: lib/shakapacker/base_strategy.rb (hunk 1 above)
Comment (ahangarha) [reply lib/shakapacker/base_strategy.rb:1]: This comment is now outdated

File: lib/tasks/webpacker/info.rake (hunk 3 above)
Comment (ahangarha) [reply lib/tasks/webpacker/info.rake]: This comment is now outdated

File: lib/webpacker/helper.rb (hunk 8 above)
Comment (ahangarha) [reply lib/webpacker/helper.rb]: This comment is now outdated

File: lib/webpacker/dev_server_runner.rb (hunk 7 above)
Comment (ahangarha) [reply lib/webpacker/dev_server_runner.rb]: This comment is now outdated

File: lib/tasks/webpacker/clean.rake (hunk 6 above)
Comment (ahangarha) [reply lib/tasks/webpacker/clean.rake]: This comment is now outdated

File: lib/tasks/webpacker/check_node.rake (hunk 2 above)
Comment (ahangarha) [reply lib/tasks/webpacker/check_node.rake]: This comment is now outdated

File: lib/tasks/shakapacker.rake (hunk 5 above)
Comment (ahangarha) [reply lib/tasks/shakapacker.rake]: Applied

File: package/rules/__tests__/swc.js [hunk 9]
```diff
@@ -38,6 +38,11 @@ const createWebpackConfig = (file, use) => {
 describe("swc", () => {
   test("process source path", async () => {
     const normalPath = `${app_javascript}/a.js`;
+    
+    console.log('ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ')
```
Comment (vaukalak) [inline package/rules/__tests__/swc.js:42]: @ahangarha is this change expected?

**Comment by vaukalak** [comment]
Otherwise LGTM

File: CHANGELOG.md [hunk 10]
```diff
@@ -8,6 +8,9 @@ Changes since last non-beta release.
 
 _Please add entries here for your pull requests that are not yet released._
 
+### Breaking changes
+- Rename Webpacker to Shakapacker in entire project including config files, binstubs, environment variables,... [PR157](https://github.com/shakacode/shakapacker/pull/157) by [ahangarha](https://github.com/ahangarha)
```
Comment (justin808) [inline CHANGELOG.md:12]: But 6.x is not a breaking change.

File: README.md [hunk 11]
```diff
@@ -154,32 +155,32 @@ yarn add @babel/core @babel/plugin-transform-runtime @babel/preset-env @babel/ru
...
-3. Provide a community supported, default webpack compilation that generates the necessary bundles and manifest, using the same configuration file. This compilation can be extended for any needs.
+2. Provide Rails view helpers, utilizing this configuration file so that a webpage can load JavaScript, CSS, and other static assets compiled by webpack, supporting bundle splitting, fingerprinting, and HMR.
+3. Provide a community-supported, default webpack compilation that generates the necessary bundles and manifest, using the same configuration file. This compilation can be extended for any needs.
 
 ## Usage
 
 ### Configuration and Code
 
-You will need your file system to correspond to the setup of your `webpacker.yml` file.
+You will need your file system to correspond to the setup of your `shakapacker.yml` file.
```
Comment (justin808) [inline README.md:179]:
```
You will need your file system to correspond to the setup of your `config/shakapacker.yml` file.
```

File: README.md [hunk 12]
```diff
@@ -210,16 +211,20 @@ Webpack intelligently includes only necessary files. In this example, the file `
...
 
-Be sure to consult the API documentation in the source code of [helper.rb](./lib/webpacker/helper.rb).
+Be sure to consult the API documentation in the source code of [helper.rb](./lib/shakapacker/helper.rb).
 
 **Note:** In order for your styles or static assets files to be available in your view, you would need to link them in your "pack" or entry file. Otherwise, Webpack won't know to package up those files.
 
 #### View Helpers `javascript_pack_tag` and `stylesheet_pack_tag`
 
-These view helpers take your `webpacker.yml` configuration file, along with the resulting webpack compilation `manifest.json` and generates the HTML to load the assets.
+These view helpers take your `shakapacker.yml` configuration file, along with the resulting webpack compilation `manifest.json` and generate the HTML to load the assets.
```
Comment (justin808) [inline README.md:227]:
```
These view helpers take your `shakapacker.yml` configuration file and the resulting webpack compilation `manifest.json` and generate the HTML to load the assets.
```

File: README.md [hunk 13]
```diff
@@ -844,14 +849,19 @@ See the doc page for [Troubleshooting](./docs/troubleshooting.md).
 
 ## Contributing
 
-We encourage you to contribute to Shakapacker/Webpacker! See [CONTRIBUTING](CONTRIBUTING.md) for guidelines about how to proceed. We have a [Slack discussion channel](https://reactrails.slack.com/join/shared_invite/enQtNjY3NTczMjczNzYxLTlmYjdiZmY3MTVlMzU2YWE0OWM0MzNiZDI0MzdkZGFiZTFkYTFkOGVjODBmOWEyYWQ3MzA2NGE1YWJjNmVlMGE).
+We encourage you to contribute to Shakapacker/Shakapacker! See [CONTRIBUTING](CONTRIBUTING.md) for guidelines about how to proceed. We have a [Slack discussion channel](https://reactrails.slack.com/join/shared_invite/enQtNjY3NTczMjczNzYxLTlmYjdiZmY3MTVlMzU2YWE0OWM0MzNiZDI0MzdkZGFiZTFkYTFkOGVjODBmOWEyYWQ3MzA2NGE1YWJjNmVlMGE).
```
Comment (justin808) [inline README.md:852]:
```
We encourage you to contribute to Shakapacker! See [CONTRIBUTING](CONTRIBUTING.md) for guidelines about how to proceed. We have a [Slack discussion channel](https://reactrails.slack.com/join/shared_invite/enQtNjY3NTczMjczNzYxLTlmYjdiZmY3MTVlMzU2YWE0OWM0MzNiZDI0MzdkZGFiZTFkYTFkOGVjODBmOWEyYWQ3MzA2NGE1YWJjNmVlMGE).
```

**Review by justin808 — CHANGES_REQUESTED** [review]
Make a branch called

webpacker-v6.7

**otherwise this PR will then reflect all the changes on the other branch**

and change base of this PR to that branch.

merge after fixes

then change other PR to be off of this new branch

then merge

then I'll push pre-release on the branch

File: CHANGELOG.md (hunk 10 above)
Comment (ahangarha) [reply CHANGELOG.md:12]: This PR is breaking. The other PR makes these changes non-breaking.

Some background:
We were going to make a breaking release (v7) out of this branch. Later we decided to add backward compatibility for perhaps the last v6.x.

File: package/rules/__tests__/swc.js (hunk 9 above)
Comment (ahangarha) [reply package/rules/__tests__/swc.js]: Fixed

File: CHANGELOG.md (hunk 10 above)
Comment (ahangarha) [reply CHANGELOG.md:12]: Done

**Review by justin808 — APPROVED** [review]

**Pull request merged**